miden-lib/src/faucets/mod.rs

```diff
@@ -0,0 +1,113 @@
+use crate::{assembler::assembler, auth::AuthScheme};
+use miden_objects::{
+    accounts::{Account, AccountCode, AccountId, AccountStorage, AccountType, AccountVault},
+    assembly::ModuleAst,
+    crypto::merkle::MerkleStore,
+    utils::{
+        string::{String, ToString},
+        vec,
+    },
+    AccountError, Felt, StarkField, Word, ZERO,
+};
+
+/// Creates a new faucet account with basic faucet interface, specified authentication scheme,
+/// and provided meta data (token symbol, decimals, max supply).
+///
+/// The basic faucet interface exposes two procedures:
+/// - `distribute`, which mints an assets and create a note for the provided recipient.
+/// - `burn`, which burns the provided asset.
+///
+/// `distribute` requires authentication. The authentication procedure is defined by the specified
+/// authentication scheme.
+/// `burn` does not require authentication and can be called by anyone.
+/// Public key information for the scheme is stored in the account storage at slot 0.
+/// The token metadata is stored in the account storage at slot 1.
+pub fn create_basic_faucet(
+    init_seed: [u8; 32],
+    symbol: String,
+    decimals: u8,
+    max_supply: Felt,
+    auth_scheme: AuthScheme,
+) -> Result<(Account, Word), AccountError> {
+    // Atm we onlt have RpoFalcon512 as authentication scheme and this is also the default in the
+    // faucet contract, so we can just use the public key as storage slot 0.
+    let auth_data: Word = match auth_scheme {
+        AuthScheme::RpoFalcon512 { pub_key } => pub_key.into(),
+    };
+
+    let account_code_src = include_str!("../../asm/faucets/basic.masm");
+    let account_code_ast = ModuleAst::parse(account_code_src)
+        .map_err(|e| AccountError::AccountCodeAssemblerError(e.into()))?;
+    let account_assembler = assembler();
+    let account_code = AccountCode::new(account_code_ast.clone(), &account_assembler)?;
+
+    // First check that the metadata is valid.
+    if decimals > 18 {
+        return Err(AccountError::FungibleFaucetInvalidMetadata(
+            "Decimals must be less than 19".to_string(),
+        ));
+    } else if symbol.len() > 3 {
+        return Err(AccountError::FungibleFaucetInvalidMetadata(
+            "Token Symbol must have exactly three characters".to_string(),
+        ));
+    } else if max_supply.as_int() == 0 {
+        return Err(AccountError::FungibleFaucetInvalidMetadata(
+            "Max supply must be > 0".to_string(),
+        ));
+    }
```

Comment:
I would create individual constants and put them at the top of the file rather than hard-coding values here. Also, a couple of other comments:

1. We need to check upper bound on the max supply. It must be smaller than $2^{63}$, I believe.
2. I would allow symbol to be longer than 3 chars (e.g., "MATIC" is 5 chars). I would probably set the limit at 6 chars. We should also probably check for zero-length strings.
3. I think 18 decimal places is too much. Max possible supply requires 19 digits to express, I believe. Maybe we set the max decimal places at 9 or 12?

Reply:
1. Since max_supply is provided as a Felt, there should be a check at Felt creation, right? I check it now again.
2. Ethereum allows for only 3 characters, right? We can do 6 characters but maybe we should try to stay as close to Ethereum as possible?
3. Also Ethereum allows for 18 decimals, rights? In order to allow for all tokens that are used somewhere else, I would do the same here tbh

Reply:
1. We need to check it because Felt range is up to $2^{64} - 2^{32} + 1$ but the max allowed supply is $2^{63} - 1$. So, something can be a valid Felt but not valid max supply.
2. From what I know, there is no limit on token symbol length as part of ERC20 standard. Do you mean something else?
3. I think we'll have to deviate here from Ethereum as Ethereum uses much bigger integers (256 bits). So, we'll need to have a strategy for translating token values anyways. For decimals, if we set the number to 18, the maximum token value we can represent about 9, which is not practical. For 12 decimal places, we can go up to about 9 million - which I think it usable (though, in most cases we'll need to go lower). Overall, I think we can set the max at 12 decimal places.